be/src/formats/parquet/page_reader.cpp

```diff
@@ -47,18 +50,40 @@ Status PageReader::next_header() {
     size_t remaining = _finish_offset - _offset;
     uint32_t header_length = 0;
 
+    RETURN_IF_ERROR(_stream->seek(_offset));
+
     do {
         allowed_page_size = std::min(std::min(allowed_page_size, remaining), kMaxPageHeaderSize);
 
         std::vector<uint8_t> page_buffer;
-        page_buffer.reserve(allowed_page_size);
-        uint8_t* page_buf = page_buffer.data();
-
-        RETURN_IF_ERROR(_stream->read_at_fully(_offset, page_buf, allowed_page_size));
+        const uint8_t* page_buf = nullptr;
+
+        // prefer peek data instead to read data.
+        bool peek_mode = false;
+        {
+            auto st = _stream->peek(allowed_page_size);
+            if (st.ok() && st.value().size() == allowed_page_size) {
+                page_buf = (const uint8_t*)st.value().data();
+                peek_mode = true;
+            } else {
+                page_buffer.reserve(allowed_page_size);
+                RETURN_IF_ERROR(_stream->read_at_fully(_offset, page_buffer.data(), allowed_page_size));
+                page_buf = page_buffer.data();
+                auto st = _stream->peek(allowed_page_size);
+                if (st.ok()) {
+                    _stats->bytes_read -= allowed_page_size;
+                    peek_mode = true;
+                }
+            }
+        }
```

Comment:
The above update involve some extra operations in the key paths. For example, we even execute `peek` function twice except the original read in many cases, which need an upper_bound for a shared buffer map. So, I not sure whether it has a performance degradation.
I even have a doubt. If we know that when AppIOReadBytes is greater than FSIOReadBytes is because the current header is read too much and some subsequent data is read from the shared buffer, then this should be understandable from our current code logic. So do we need to make these changes to correct this statistic?

Reply:
> The above update involve some extra operations in the key paths. For example, we even execute peek function twice except the original read in many cases, which need an upper_bound for a shared buffer map. So, I not sure whether it has a performance degradation.

Peek is a very light operation. In shared buffer, it's just a map look up key. Comparing to IO operation, this operation is quite lightweight.

> I even have a doubt. If we know that when AppIOReadBytes is greater than FSIOReadBytes is because the current header is read too much and some subsequent data is read from the shared buffer, then this should be understandable from our current code logic. So do we need to make these changes to correct this statistic?

I think to some extent, it's necessary if we still show `AppIOReadBytes` this metric.